File: glug-model/src/test/java/glug/model/ThreadedSystemTest.java
package glug.model;

import glug.model.time.LogInstant;
import glug.model.time.LogInterval;
import org.junit.Before;
import org.junit.Test;
import org.threeten.extra.Interval;

import java.util.Map;

import static com.google.common.collect.ImmutableMap.of;
import static java.time.Duration.ofSeconds;
import static java.time.Instant.ofEpochMilli;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.Matchers.nullValue;


public class ThreadedSystemTest {
    private ThreadedSystem threadedSystem;

    @Before
    public void setUp() {
        threadedSystem = new ThreadedSystem();
    }

    @Test
    public void shouldReturnNullWhenReportingIntervalCoveredByEmptyThreadsRatherThanThrowNullPointer() {
        threadedSystem.getOrCreateThread("Thread without sigint");

        assertThat(threadedSystem.getIntervalCoveredByAllThreads(), nullValue());
    }

    @Test
    public void shouldUnderstandThatThreadsAreDifferentDammit() {
        threadedSystem.getOrCreateThread("Timeout guard");
        threadedSystem.getOrCreateThread("timerFactory");
        assertThat(threadedSystem.getThreads().size(), equalTo(2));
    }

    @Test
    public void shouldCountStuff() {
        IntervalTypeDescriptor intervalType = null;
        LogInterval logInterval = new LogInterval(ofSeconds(10), new LogInstant(15));
        //threadedSystem.getOrCreateThread("a").add(significantInterval)
        Map<Object, Integer> countMap = threadedSystem.countOccurencesDuring(logInterval, intervalType);
    }

    @Test
    public void shouldHandleGettingTotalIntervalEvenIfSomeThreadsHaveNoIntervalData() {
        com.madgag.interval.Interval<LogInstant> interval = new LogInterval(Interval.of(ofEpochMilli(3000), ofEpochMilli(7000)));
        threadedSystem.getOrCreateThread("A1");
        threadedSystem.getOrCreateThread("B2").add(new SignificantInterval(of("type", "My Type"), interval));
        threadedSystem.getOrCreateThread("C3");
        assertThat(threadedSystem.getIntervalCoveredByAllThreads(), equalTo(interval));
    }
}

File: glug-model/src/test/scala/glug/model/ThreadedSystemTest.scala
package glug.model

import java.util.Collections.emptyMap

import glug.model.time.{LogInstant, LogInterval}
import org.scalatest.{FlatSpec, Matchers}

class ThreadedSystemTest extends FlatSpec with Matchers {

  "ThreadedSystem" should "not error when reporting interval covered by empty threads" in {
    val threadedSystem = new ThreadedSystem
    threadedSystem.getOrCreateThread("Thread without sigint")
    threadedSystem.getIntervalCoveredByAllThreads
  }

  it should "understand that threads are different dammit" in {
    val threadedSystem = new ThreadedSystem
    threadedSystem.getOrCreateThread("Timeout guard")
    threadedSystem.getOrCreateThread("timerFactory")
    threadedSystem.getNumThreads shouldBe 2
  }

  it should "handle getting total interval even if some threads have no interval data" in {
    val threadedSystem = new ThreadedSystem
    val interval = LogInterval(LogInstant(3000L), LogInstant(7000L))
    threadedSystem.getOrCreateThread("A1")
    threadedSystem.getOrCreateThread("B2").add(new SignificantInterval(emptyMap(), interval))
    threadedSystem.getOrCreateThread("C3")
    threadedSystem.getIntervalCoveredByAllThreads shouldBe interval
  }

}